File: C++/Longest_palindromic_substring.cpp
#include <iostream>
#include <vector>

class LongestPalindromicSubstring {
public:
    std::string longestPalindrome(std::string s) {
        if (s.empty()) return "";
        
        int n = s.size();
        int start = 0, maxLength = 1;
        std::vector<std::vector<bool>> dp(n, std::vector<bool>(n, false));

        for (int i = 0; i < n; ++i)
            dp[i][i] = true;

        for (int i = 0; i < n - 1; ++i) {
            if (s[i] == s[i + 1]) {
                dp[i][i + 1] = true;
                start = i;
                maxLength = 2;
            }
        }

        for (int len = 3; len <= n; ++len) {
            for (int i = 0; i < n - len + 1; ++i) {
                int j = i + len - 1;
                if (s[i] == s[j] && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    start = i;
                    maxLength = len;
                }
            }
        }

        return s.substr(start, maxLength);
    }
};

int main() {
    LongestPalindromicSubstring lps;
    std::string s = "babad";
    std::cout << "Longest palindromic substring of " << s << ": " << lps.longestPalindrome(s) << "\n";

    return 0;
}

File: Python/Longest_palindromic_substring.py
def longest_palindromic_substring(s):
    def expand_around_center(left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left + 1:right]
    
    longest = ""
    for i in range(len(s)):
        odd_palindrome = expand_around_center(i, i)
        even_palindrome = expand_around_center(i, i + 1)
        longest = max(longest, odd_palindrome, even_palindrome, key=len)
    return longest

if __name__ == "__main__":
    s = input("Enter a string: ")
    print(f"Longest palindromic substring: {longest_palindromic_substring(s)}")
